File: dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/boot/internal/XMLUtils.java
/*******************************************************************************
 * Copyright (c) 2012 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package com.ibm.ws.kernel.boot.internal;

import java.lang.reflect.Field;
import java.util.Set;

import javax.xml.stream.XMLStreamReader;

/**
 * Using the class to process XML files
 */
public class XMLUtils {
    /**
     * Get the element attribute value
     * 
     * @param reader
     * @param localName
     * @return
     */
    public static String getAttribute(XMLStreamReader reader, String localName) {
        int count = reader.getAttributeCount();
        for (int i = 0; i < count; i++) {
            String name = reader.getAttributeLocalName(i);
            if (localName.equals(name)) {
                return reader.getAttributeValue(i);
            }
        }
        return null;
    }

    /**
     * Create the instance by parse the element, the instance's class must have the empty construct.
     * The clazz must have the fields in attrNames and all the fields type must be String
     * 
     * @param reader
     * @param clazz
     * @param attrNames
     * @return
     */
    public static <T> T createInstanceByElement(XMLStreamReader reader, Class<T> clazz, Set<String> attrNames) {
        if (reader == null || clazz == null || attrNames == null)
            return null;

        try {
            T instance = clazz.newInstance();

            int count = reader.getAttributeCount();
            int matchCount = attrNames.size();

            for (int i = 0; i < count && matchCount > 0; ++i) {
                String name = reader.getAttributeLocalName(i);
                String value = reader.getAttributeValue(i);

                if (attrNames.contains(name)) {
                    Field field = clazz.getDeclaredField(name);
                    field.setAccessible(true);
                    field.set(instance, value);
                    matchCount--;
                }
            }
            return instance;
        } catch (InstantiationException e) {
            throw new RuntimeException(e);
        } catch (IllegalAccessException e) {
            throw new RuntimeException(e);
        } catch (SecurityException e) {
            throw new RuntimeException(e);
        } catch (NoSuchFieldException e) {
            throw new RuntimeException(e);
        }
    }
}

File: dev/com.ibm.ws.ui.tool.serverConfig/resources/WEB-CONTENT/js/utils/xmlUtils.js
/*******************************************************************************
 * Copyright (c) 2015 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/

var xmlUtils = (function() {
    "use strict";
    
    var getElementPath = function(element) {
        var path = [];
        while(element.parentNode !== null && element.parentNode !== undefined) {
            path.unshift(element.nodeName);
            element = element.parentNode;
        } 
        return path;
    };
    
    
    var elementHasChildren = function(element) {
        for(var child = element.firstChild; child !== null && child !== undefined; child = child.nextSibling) {
            if(child.nodeType === 1) {
                return true;
            }
        }
        return false;
    };
    
    
    var getElementChildrenCount = function(element) {
        var count = 0;
        for(var child = element.firstChild; child !== null && child !== undefined; child = child.nextSibling) {
            if(child.nodeType === 1) {
                count++;
            }
        }
        return count;
    };
    
    
    var getChildElementAtIndex = function(parent, index) {
        var currentIndex = 0;
        for(var child = parent.firstChild; child !== null && child !== undefined; child = child.nextSibling) {
            if(child.nodeType === 1) {
                if(currentIndex === index) {
                    return child;
                } else {
                    currentIndex++;
                }
            }
        }
        return null;
    };
    
    
    var getElementIndex = function(element) {
        var index = 0;
        for(var child = element.parentNode.firstChild; child !== null && child !== undefined; child = child.nextSibling) {
            if(child.nodeType === 1) {
                if(child === element) {
                    return index;
                } else {
                    index++;
                }
            }
        }
        return null;
    };
    
    
    var canMoveUp = function(element) {
        for(var child = element.parentNode.firstChild; child !== null && child !== undefined; child = child.nextSibling) {
            if(child.nodeType === 1) {
                return child !== element;
            }
        }
        return true;
    };
    
    
    var canMoveDown = function(element) {
        for(var child = element.parentNode.lastChild; child !== null && child !== undefined; child = child.previousSibling) {
            if(child.nodeType === 1) {
                return child !== element;
            }
        }
        return true;
    };
    
    
    var getElementNestLevel = function(element) {
        var level = 0;
        while(element.parentNode !== null && element.parentNode !== undefined) {
            element = element.parentNode;
            level++;
        }
        return level;
    };
    
    
    var getDocumentDeclaration = function(documentText) {
        var documentDeclaration = null;
        if(documentText.length > 4 && documentText.substring(0, 2) === "<?") {
            return documentText.substring(0, documentText.indexOf("?>") + 2);
        }
        return documentDeclaration;
    };
    
    
    return {
        getElementPath: getElementPath,
        elementHasChildren: elementHasChildren,
        getElementChildrenCount: getElementChildrenCount,
        getChildElementAtIndex: getChildElementAtIndex,
        getElementIndex: getElementIndex,
        canMoveUp: canMoveUp,
        canMoveDown: canMoveDown,
        getElementNestLevel: getElementNestLevel,
        getDocumentDeclaration: getDocumentDeclaration
    };
    
})();
